### PR TITLE
Add option of launching CIV in GVT-d mode

### PR DESCRIPTION
**Pull request by lyintel**
Include setup and start script.

### CODE REVIEW

File: groups/device-specific/caas/setup_host.sh [hunk 1]
```diff
@@ -42,6 +42,30 @@ function ubu_install_qemu(){
...
+	    --enable-kvm \
+	    --disable-xen \
+	    --enable-libusb \
+	    --enable-debug-info \
+	    --enable-debug \
+	    --enable-vhost-net \
+	    --disable-debug-tcg \
+	    --target-list=x86_64-softmmu \
+	    --audio-drv-list=pa
+	 make -j24
```
Comment (YadongQi) [inline groups/device-specific/caas/setup_host.sh:64]: please be aware of the indent.

File: groups/device-specific/caas/setup_host.sh [hunk 2]
```diff
@@ -149,19 +197,43 @@ function save_env(){
 	fi
 }
 
+function install_service(){
```
Comment (YadongQi) [inline groups/device-specific/caas/setup_host.sh:200]: name it with "install_auto_start_service" ?

File: groups/device-specific/caas/setup_host.sh [hunk 3]
```diff
@@ -149,19 +197,43 @@ function save_env(){
...
+	check_kernel
+    install_service
+	if [[ $1 == "--gvtd" ]]; then
+		ubu_install_qemu_gvtd
+		ubu_build_ovmf_gvtd
+		ubu_enable_host_gvtd
+	else
+	    ubu_install_qemu
+	    ubu_build_ovmf
+	    ubu_enable_host_gvtg
```
Comment (YadongQi) [inline groups/device-specific/caas/setup_host.sh:231]: use tab as indent.

File: groups/device-specific/caas/start_android_qcow2.sh [hunk 4]
```diff
@@ -20,6 +20,20 @@ function setup_vgpu(){
...
+function setup_vgpu_gvtd(){
+	res=0
+	if [ ! -d $GVTg_DEV_PATH/$GVTg_VGPU_UUID ]; then
+		echo "Creating VGPU..."
+	    sudo sh -c "modprobe vfio"
+	    sudo sh -c "modprobe vfio_pci"
+	    intelVGA_id=`sudo sh -c "lspci -D -nn |grep '02.0 VGA' |grep -o 8086:[0-9a-f][0-9a-f][0-9a-f][0-9a-f]"`
+	    sudo sh -c "echo 0000:00:02.0 > /sys/bus/pci/devices/0000:00:02.0/driver/unbind"
+	    sudo sh -c "echo ${intelVGA_id/:/ } > /sys/bus/pci/drivers/vfio-pci/new_id"
+		res=$?
```
Comment (YadongQi) [inline groups/device-specific/caas/start_android_qcow2.sh:32]: indent

File: groups/device-specific/caas/start_android_qcow2.sh [hunk 5]
```diff
@@ -153,6 +167,12 @@ function launch_hwrender(){
 	fi
 }
 
+function launch_hwrender_gvtd(){
+	qemu-system-x86_64 \
+          -device vfio-pci,host=00:02.0,x-igd-gms=2,id=hostdev0,bus=pcie.0,addr=0x2,x-igd-opregion=on \
+	  $common_options
```
Comment (YadongQi) [inline groups/device-specific/caas/start_android_qcow2.sh:173]: indent

**Comment by lyintel** [comment]
All the indent issues are fixed, and the function name changed to "install_auto_start_service"

File: groups/device-specific/caas/setup_host.sh [hunk 6]
```diff
@@ -23,20 +23,44 @@ function ubu_install_qemu(){
...
+	cd ../
+}
+
+function ubu_install_qemu_gvtd(){
+	apt purge -y "qemu*"
+	apt autoremove -y
+	apt install -y git python-dev libfdt-dev libpixman-1-dev libssl-dev socat autoconf libtool uml-utilities bridge-utils liblzma-dev libc6-dev libegl1-mesa-dev libepoxy-dev libdrm-dev libgbm-dev libaio-dev libusb-1.0.0-dev bison flex gcc g++ flex pkg-config python-pip libpulse-dev uuid-runtime uuid 
+	wget https://download.qemu.org/$QEMU_REL.tar.xz
+	tar -xf $QEMU_REL.tar.xz
+	mv 0001-Revert-Revert-vfio-pci-quirks.c-Disable-stolen-memor.patch $QEMU_REL/
```
Comment (YadongQi) [inline groups/device-specific/caas/setup_host.sh:51]: could you use wget/curl to download patch here? then there is no need to download patch mannually

File: groups/device-specific/caas/setup_host.sh [hunk 7]
```diff
@@ -60,6 +84,18 @@ function ubu_build_ovmf(){
 	cd ../../../
 }
 
+function ubu_build_ovmf_gvtd(){
+	sudo apt install -y uuid-dev nasm acpidump iasl
+	mv OvmfPkg-add-IgdAssgingmentDxe-for-qemu-4_2_0.patch $QEMU_REL/roms/edk2/
```
Comment (YadongQi) [inline groups/device-specific/caas/setup_host.sh:89]: download patch here automatically by wget/curl

File: groups/device-specific/caas/setup_host.sh [hunk 8]
```diff
@@ -149,19 +197,43 @@ function save_env(){
...
 	check_network
 	ubu_changes_require
-	ubu_install_qemu
-	ubu_build_ovmf
-	ubu_enable_host_gvtg
-	get_required_scripts
-	check_kernel
 	save_env
+	check_kernel
+	install_auto_start_service
```
Comment (YadongQi) [inline groups/device-specific/caas/setup_host.sh:223]: Currently, do not make it auto start by default. We may add an option later.

File: groups/device-specific/caas/setup_host.sh [hunk 9]
```diff
@@ -60,6 +84,18 @@ function ubu_build_ovmf(){
 	cd ../../../
 }
 
+function ubu_build_ovmf_gvtd(){
```
Comment (YadongQi) [inline groups/device-specific/caas/setup_host.sh:87]: seems the only difference between ubu_build_ovmf() and ubu_build_ovmf_gvtd() is the patching work. Could you unify them into 1?

File: groups/device-specific/caas/setup_host.sh [hunk 10]
```diff
@@ -23,20 +23,44 @@ function ubu_install_qemu(){
...
+		--enable-gtk \
+		--enable-virtfs \
+		--target-list=x86_64-softmmu \
+		--audio-drv-list=pa
+	make -j24
+	make install
+	cd ../
+}
+
+function ubu_install_qemu_gvtd(){
```
Comment (YadongQi) [inline groups/device-specific/caas/setup_host.sh:45]: could you unify ubu_install_qemu_gvtd() and ubu_install_qemu() ?

File: groups/device-specific/caas/0001-Revert-Revert-vfio-pci-quirks.c-Disable-stolen-memor.patch [hunk 11]
```diff
@@ -0,0 +1,124 @@
+From e9cb22a65479e7f9730198cf39ed979aa58b5689 Mon Sep 17 00:00:00 2001
+From: lyintel <yang.a.lu@intel.com>
+Date: Wed, 4 Mar 2020 15:19:57 +0800
+Subject: [PATCH 1/2] Revert "Revert "vfio/pci-quirks.c: Disable stolen memory
+ for igd VFIO""
```
Comment (JeevakaPrabu) [inline groups/device-specific/caas/0001-Revert-Revert-vfio-pci-quirks.c-Disable-stolen-memor.patch:5]: please dont keep the patches under mixins folder. Better create folder under vendor/intel/utils like
vendor/intel/utils/host/qemu/
vendor/intel/utils/host/ovmf/

Update the setup_host.sh to download this patches and apply it before building Qemu and OVMF

File: groups/device-specific/caas/setup_host.sh [hunk 12]
```diff
@@ -149,19 +197,43 @@ function save_env(){
...
-	ubu_install_qemu
-	ubu_build_ovmf
-	ubu_enable_host_gvtg
-	get_required_scripts
-	check_kernel
 	save_env
+	check_kernel
+	install_auto_start_service
+	if [[ $1 == "--gvtd" ]]; then
+		ubu_install_qemu_gvtd
```
Comment (JeevakaPrabu) [inline groups/device-specific/caas/setup_host.sh:225]: As we need to update mesa for some platforms, can you provide an option to clone and update mesa as part of this script itself?

File: groups/device-specific/caas/setup_host.sh (hunk 1 above)
Comment (lyintel) [reply groups/device-specific/caas/setup_host.sh]: resolved.

File: groups/device-specific/caas/setup_host.sh (hunk 12 above)
Comment (lyintel) [reply groups/device-specific/caas/setup_host.sh:225]: The mesa update option will be discussed in future, this patch will only add the option of launching GVT-d

File: groups/device-specific/caas/setup_host.sh (hunk 6 above)
Comment (lyintel) [reply groups/device-specific/caas/setup_host.sh:51]: Patch will be move to proper folder in future once we decide where to put them in.

File: groups/device-specific/caas/setup_host.sh (hunk 7 above)
Comment (lyintel) [reply groups/device-specific/caas/setup_host.sh:89]: Patch will be move to proper folder in future once we decide where to put them in. Then download patch here automatically by wget/curl.

File: groups/device-specific/caas/setup_host.sh (hunk 8 above)
Comment (lyintel) [reply groups/device-specific/caas/setup_host.sh]: Done

File: groups/device-specific/caas/setup_host.sh (hunk 9 above)
Comment (lyintel) [reply groups/device-specific/caas/setup_host.sh:87]: There is also a build parameter difference, prefer to keep them into 2.

File: groups/device-specific/caas/setup_host.sh (hunk 10 above)
Comment (lyintel) [reply groups/device-specific/caas/setup_host.sh:45]: There is also a build parameter difference, prefer to keep them into 2.

File: groups/device-specific/caas/0001-Revert-Revert-vfio-pci-quirks.c-Disable-stolen-memor.patch (hunk 11 above)
Comment (lyintel) [reply groups/device-specific/caas/0001-Revert-Revert-vfio-pci-quirks.c-Disable-stolen-memor.patch:5]: Patch will be move to proper folder in future once we decide where to put them in. Then download patch here automatically by wget/curl.

**Comment by lyintel** [comment]
Tested this PR on KBL NUC, and the gvtd option works well, will continue to merge the PR.

File: groups/device-specific/caas/start_android_qcow2.sh [hunk 13]
```diff
@@ -26,6 +26,20 @@ function setup_vgpu(){
 	return $res
 }
 
+function setup_vgpu_gvtd(){
```
Comment (YuanjunHuang) [inline groups/device-specific/caas/start_android_qcow2.sh:29]: the naming here seems confusing. For gvtd, it is not vgpu - virtual gpu. Suggest name to setup_gvtd.

File: groups/device-specific/caas/start_android_qcow2.sh [hunk 14]
```diff
@@ -26,6 +26,20 @@ function setup_vgpu(){
...
+	res=0
+	if [ ! -d $GVTg_DEV_PATH/$GVTg_VGPU_UUID ]; then
+		echo "Creating VGPU..."
```
Comment (YuanjunHuang) [inline groups/device-specific/caas/start_android_qcow2.sh:32]: Suggest to change to 'Unbinding GPU to the host and binding dedicated GPU to the guest ...'

File: groups/device-specific/caas/start_android_qcow2.sh [hunk 15]
```diff
@@ -217,10 +237,18 @@ vno=$(echo $version | \
...
 	network_setup
 	if [[ $? == 0 ]]; then
-		launch_hwrender $1
+		if [[ $2 == "--gvtd" ]]; then
+			launch_hwrender_gvtd
+		else
+			launch_hwrender $1
+		fi
 	else
 		echo "W: Failed to create vgpu, fall to software rendering"
```
Comment (YuanjunHuang) [inline groups/device-specific/caas/start_android_qcow2.sh:253]: Suggest to change accordingly to "W: Failed to create vgpu or bind dedicated GPU to the guest, fall to software rendering"

File: groups/device-specific/caas/0001-Revert-Revert-vfio-pci-quirks.c-Disable-stolen-memor.patch (hunk 11 above)
Comment (lyintel) [reply groups/device-specific/caas/0001-Revert-Revert-vfio-pci-quirks.c-Disable-stolen-memor.patch]: The patches of QEMU and OVMF have been moved to vendor/intel/utils/.

File: groups/device-specific/caas/start_android_qcow2.sh (hunk 13 above)
Comment (lyintel) [reply groups/device-specific/caas/start_android_qcow2.sh]: function name and related output message changed, thank you

**Pull request merged**